cis194/week4/zhansongl/Wholemeal.hs

```diff
@@ -0,0 +1,71 @@
+module Wholemeal where
+
+import Data.Bool (bool)
+import qualified Data.Array as A
+
+fun1 :: [Integer] -> Integer
+fun1 [] = 1
+fun1 (x:xs)
+  | even x    = (x - 2) * fun1 xs
+  | otherwise = fun1 xs
+
+fun1' :: [Integer] -> Integer
+fun1' = foldr (*) 1 . map (flip (-) 2) . filter even
+
+fun2 :: Integer -> Integer
+fun2 1 = 0
+fun2 n | even n    = n + fun2 (n `div` 2)
+       | otherwise = fun2 (3 * n + 1)
+
+fun2' :: Integer -> Integer
+fun2' = sum . filter even . takeWhile (/= 1) . iterate next
+  where next n = bool (3 * n + 1) (n `div` 2) $ even n
+
+
+data Tree a = Leaf
+            | Node Integer (Tree a) a (Tree a)
+  deriving (Show, Eq)
+
+getHeight :: Tree a -> Integer
+getHeight Leaf = (-1)
+getHeight (Node h _ _ _) = h
+
+insert :: a -> Tree a -> Tree a
+insert x Leaf = Node 0 Leaf x Leaf
+insert x (Node h left y right)
+  | getHeight left < getHeight right =
+      let h' = 1 + max (getHeight left') (getHeight right)
+          left' = insert x left
+      in  Node h' left' y right
+  | otherwise =
+      let h' = 1 + max (getHeight left) (getHeight right')
+          right' = insert x right
+      in  Node h' left y right'
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insert Leaf
+
+xor :: [Bool] -> Bool
+xor = odd . length . filter id
+
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\x ys -> (f x):ys) []
+
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f base = foldr (flip f) base . reverse
```

Comment:
This is correct logic.
Now let's make it more efficient. Because `reverse` is doing a `folding` on the list. Plus `foldr`, we are folding twice here. Is there a way only folding once? (hint: how about folding out a recursive function then call with base?) The correct answer is the implementation of `foldl'`

Reply:
fixed, took me a while to figure this one out